hazelcast/src/main/java/com/hazelcast/internal/blackbox/impl/BlackboxImpl.java

```diff
@@ -0,0 +1,315 @@
+import com.hazelcast.internal.blackbox.LongSensorInput;
+import com.hazelcast.internal.blackbox.Sensor;
+import com.hazelcast.internal.blackbox.SensorView;
+import com.hazelcast.internal.blackbox.sensorpacks.ClassLoadingSensorPack;
+import com.hazelcast.internal.blackbox.sensorpacks.GarbageCollectionSensorPack;
+import com.hazelcast.internal.blackbox.sensorpacks.OperatingSystemSensorPack;
+import com.hazelcast.internal.blackbox.sensorpacks.RuntimeSensorPack;
+import com.hazelcast.internal.blackbox.sensorpacks.ThreadSensorPack;
+import com.hazelcast.logging.ILogger;
+
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.hazelcast.util.ValidationUtil.checkNotNull;
+import static java.lang.String.format;
+
+/**
+ * The {@link Blackbox} implementation.
+ */
+public class BlackboxImpl implements Blackbox {
+
+    private static final int STRIPED_LOCK_SIZE = 100;
+
+    private final ConcurrentMap<String, SensorImpl> sensors = new ConcurrentHashMap<String, SensorImpl>();
+    private final Object[] stripedLock = new Object[STRIPED_LOCK_SIZE];
+    private final ScheduledExecutorService scheduledExecutorService = new ScheduledThreadPoolExecutor(2);
+    private final ILogger logger;
+    private final SensorInputAnnotationScanner annotationScanner = new SensorInputAnnotationScanner(this);
+    private final AtomicInteger modCount = new AtomicInteger();
+
+    public BlackboxImpl(ILogger logger) {
+        this.logger = logger;
+
+        for (int k = 0; k < stripedLock.length; k++) {
+            stripedLock[k] = new Object();
+        }
+
+
+        RuntimeSensorPack.register(this);
+        GarbageCollectionSensorPack.register(this);
+        OperatingSystemSensorPack.register(this);
+        ThreadSensorPack.register(this);
+        ClassLoadingSensorPack.register(this);
+    }
+
+    @Override
+    public int modCount() {
+        return modCount.get();
+    }
+
+    public void shutdown() {
+        scheduledExecutorService.shutdown();
+    }
+
+    @Override
+    public void unregister(Object source) {
+        checkNotNull(source, "source can't be null");
+
+        for (Map.Entry<String, SensorImpl> entry : sensors.entrySet()) {
+            String parameter = entry.getKey();
+            SensorImpl sensor = entry.getValue();
+            synchronized (getLock(parameter)) {
+                if (sensor.source == source) {
+                    sensor.destroy();
+                }
+            }
+        }
+    }
+
+    @Override
+    public void register(String parameterPrefix, Object source) {
```

Comment:
This is an example of a dangerous overload whose signature subsumes other overloads with more specialized parameter types. For sanity the method should type-check and delegate to the specializations. Another, probably better, choice is to eliminate overloading by renaming.